File: octave/quad_adapt_test.m
% TEST: Numeric Integration using adaptive Trapezoidal method

f = inline('-2*x^(-3) * cos(x^(-2))');
a = 0.5;
b = 100;
Iexp = sin(10^(-4)) - sin(4) % Expected integral value

#f = inline('0.2 + 25*x - 200*x^2 + 675*x^3 - 900*x^4 + 400*x^5');
#a = 0.0;
#b = 0.8;
#Iexp = 1.640533    % Expected integral value

N = 10000;    % Number of points for the smooth plot

%------------------------------------------------------------------------------

% Trapezoid adaptive method with relative error
display('> Trapezoid adaptive method');
[ I x ] = quad_trapezoid_adapt(a, b, f, 10^(-4));
n = length(x)
I
E = abs(Iexp - I)/Iexp   % relative error

% Trapezoid method with relative error
display('> Trapezoid method');
I = quad_trapezoid(a, b, f, n)
E = abs(Iexp - I)/Iexp   % relative error

% 1/3 Simpson's method with relative error
display('> Simpson method');
I = quad_simpson(a, b, f, n)
E = abs(Iexp - I)/Iexp

%------------------------------------------------------------------------------

% Simpson adaptive method with relative error
display('> Simpson adaptive method');
[ I x ] = quad_simpson_adapt(a, b, f, 10^(-4));
n = length(x)
I
E = abs(Iexp - I)/Iexp   % relative error

% Trapezoid method with relative error
display('> Trapezoid method');
I = quad_trapezoid(a, b, f, n)
E = abs(Iexp - I)/Iexp   % relative error

% 1/3 Simpson's method with relative error
display('> Simpson method');
I = quad_simpson(a, b, f, n)
E = abs(Iexp - I)/Iexp


#set(gca, 'XScale', 'log');
hold on
X = linspace(a, b, N);
Y = zeros(1, N);
for i = 1:N
    Y(i) = f(X(i));
endfor
plot(X, Y, 'b');
y = zeros(1,n);
for i = 1:n
    y(i) = f(x(i));
#    plot([x(i) x(i)], [0 y(i)], 'k');
endfor
plot(x, y, '+k')
hold off

% Wait for input
%pause;

File: python/quad_adapt_test.py
# TEST: Numeric Integration using adaptive Trapezoidal and Simpson methods

from quad import quad_trapezoid, quad_simpson
from quad import quad_trapezoid_adapt, quad_simpson_adapt
import numpy as np
import matplotlib.pyplot as plt


f = lambda x: -2*x**(-3) * np.cos(x**(-2))
a = 0.5;
b = 100;
Iexp = np.sin(10**(-4)) - np.sin(4) # Expected integral value

tol = 10**(-4)  # Error tolerance

print('---------------------------------------')
# Trapezoid adaprive method with relative error
I, x = quad_trapezoid_adapt(a, b, f, tol)
E = abs(Iexp - I)/Iexp      # Relative error
n = len(x)
print('> Trapezoid adaptive method\n I = {}\n E = {}\n n = {}'.format(I, E, n))
# Trapezoid method with relative error
I = quad_trapezoid(a, b, f, n-1)
E = abs(Iexp - I)/Iexp      # Relative error
print('> Trapezoid method\n I = {}\n E = {}\n n = {}'.format(I, E, n))
# Simpson's method with relative error
I = quad_simpson(a, b, f, n-1)
E = abs(Iexp - I)/Iexp      # Relative error
print('> Simpson method\n I = {}\n E = {}\n n = {}'.format(I, E, n))

print('---------------------------------------')
# Trapezoid adaprive method with relative error
I, x = quad_simpson_adapt(a, b, f, tol)
E = abs(Iexp - I)/Iexp      # Relative error
n = len(x)
print('> Simpson adaptive method\n I = {}\n E = {}\n n = {}'.format(I, E, n))
# Trapezoid method with relative error
I = quad_trapezoid(a, b, f, n-1)
E = abs(Iexp - I)/Iexp      # Relative error
print('> Trapezoid method\n I = {}\n E = {}\n n = {}'.format(I, E, n))
# Simpson's method with relative error
I = quad_simpson(a, b, f, n-1)
E = abs(Iexp - I)/Iexp      # Relative error
print('> Simpson method\n I = {}\n E = {}\n n = {}'.format(I, E, n))

# Plot function graph using the last Simpson nodes
N = 10000    # Number of points for a smooth plot
X = np.linspace(a, b, N)
#plt.plot(X, f(X))
fig, ax = plt.subplots()
ax.semilogx(X, f(X))

y = [ f(xi) for xi in x ]
for xi in x:
    ax.semilogx([xi,xi], [0, f(xi)],
             color='black', linestyle='dashed', linewidth=1)
plt.scatter(x, y, linewidths=0.2, marker='+')
plt.grid(True)
plt.show();
